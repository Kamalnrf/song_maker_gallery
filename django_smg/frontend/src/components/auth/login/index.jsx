import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import {
  login as loginAction,
  clearError,
  tryToken,
} from "../../../actions/auth.action";
import { getUserGalleries } from "../../../actions/user";
import { Link, Redirect } from "react-router-dom";
import UsernamePassword from "./username_password";
import { windowLocation } from "../../../util/window";
import { Button } from "../../common/styles";

import { ErrorHandler } from "./error_handler";

const NeedAccount = styled(Button)`
  min-width: 14px;
  background-color: #94c732;
`;

const ForgotPass = styled(Button)`
  min-width: 14px;
  background-color: #00bcd1;
`;

const login = (props) => {
  const [username, updateUsername] = useState("");
  const [password, updatePassword] = useState("");

  const token = localStorage.getItem("token");

  useEffect(() => {
    if (token) {
      props.tryToken(token);
    }
  }, []);

  if (props.isAuthenticated) {
    return <Redirect to="/teacher" />;
  }

  const submit = (e) => {
    e.preventDefault();
    props.loginAction({ username: username, password: password });
  };

  return (
    <div>
      <ErrorHandler err={props.authError} />
      <h1>login!</h1>
      <div className="description">
        <UsernamePassword
          username={(u) => updateUsername(u)}
          password={(p) => updatePassword(p)}
          submit={submit}
        />
      </div>
      <div>
        <Link to="/signup">
          <NeedAccount>Need an account?</NeedAccount>
        </Link>
        <a href={`${windowLocation("origin")}/accounts/password_reset/`}>
          <ForgotPass>Forgot your password?</ForgotPass>
        </a>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    authError: state.auth.authError,
    isAuthenticated: state.auth.isAuthenticated,
    galleries: state.user.galleries,
  };
};

export default connect(mapStateToProps, {
  loginAction,
  clearError,
  getUserGalleries,
  tryToken,
})(login);
