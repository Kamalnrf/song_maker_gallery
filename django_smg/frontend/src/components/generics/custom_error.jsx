import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import { Blanket } from "./styles";

export const StyledCustomError = styled(Blanket)`
  & > div {
    width: 70%;
    margin: auto;
  }
`;

/**
 * Generic error modal that renders props.children in the modal window
 */
export const CustomError = ({ children, onOk }) => {
  return (
    <StyledCustomError>
      <div>
        {children}
        <button
          data-testid="onOkButton"
          style={{ backgroundColor: "#4caf50" }}
          onClick={onOk}
        >
          Ok
        </button>
      </div>
    </StyledCustomError>
  );
};

CustomError.propTypes = {
  children: PropTypes.node.isRequired,
  onOk: PropTypes.func.isRequired,
};
/**
 * Render an array of error messages
 */
export const ErrorArray = (props) => {
  return (
    <CustomError onOk={() => props.onOk()}>
      <h2 data-testid="customError header">{props.header}</h2>
      {props.message.map((par, i) => (
        <p key={i}>{par}</p>
      ))}
    </CustomError>
  );
};

ErrorArray.propTypes = {
  header: PropTypes.string,
  message: PropTypes.array.isRequired,
  onOk: PropTypes.func.isRequired,
};

ErrorArray.defaultProps = {
  header: "Error",
};
