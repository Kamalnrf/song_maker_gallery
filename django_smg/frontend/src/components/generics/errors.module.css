.container {
  width: 70%;
  margin: auto;
}

.container > p,
ul,
li {
  text-align: justify;
}

.container > ul {
  margin-left: 40px;
}
