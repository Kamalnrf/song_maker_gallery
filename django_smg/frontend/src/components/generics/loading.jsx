import React from "react";
import styles from "./loading.module.css";

const loading = (props) => {
  return (
    <div className={styles.lds_ring}>
      <div></div>
      <div></div>
      <div></div>
      <div></div>
    </div>
  );
};

export default loading;
