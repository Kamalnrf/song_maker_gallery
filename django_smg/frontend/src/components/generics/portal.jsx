import { useState, useEffect } from "react";
import ReactDOM from "react-dom";

/* Makes an element that jest / react-testing library can test against */
const getElement = () => {
  const el = document.createElement("div");
  el.setAttribute("data-testid", "portalContainer");
  el.id = "portalContainer";
  return el;
};

/**
 * Very simple portal that mounts children in a separate node from the rest
 * of the app, and destroys that dom node on dismount.
 */
export const Portal = ({ children }) => {
  const [modalEl, setModalEl] = useState(null);
  useEffect(() => {
    const el = getElement();
    document.body.appendChild(el);
    setModalEl(el);
    return () => {
      document.body.removeChild(document.getElementById("portalContainer"));
    };
  }, []);
  return modalEl ? ReactDOM.createPortal(children, modalEl) : null;
};
