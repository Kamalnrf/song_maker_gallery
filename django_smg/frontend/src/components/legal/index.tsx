/**
 * terms of service
 */

import React, { useState, Fragment } from "react";

import { TosText } from "./terms_of_service";
import { PrivacyText } from "./privacy_policy";

export { TosPage } from "./terms_of_service";
export { PrivacyPage } from "./privacy_policy";

export const Tos = () => {
  const [blanket, setBlanket] = useState(null);
  return (
    <Fragment>
      {blanket}
      <a onClick={() => setBlanket(<TosText onOk={() => setBlanket(null)} />)}>
        Terms of Service
      </a>
    </Fragment>
  );
};

export const Privacy = (props) => {
  const [blanket, setBlanket] = useState(null);
  return (
    <Fragment>
      {blanket}
      <a
        onClick={() =>
          setBlanket(<PrivacyText onOk={() => setBlanket(null)} />)
        }
      >
        Privacy Policy
      </a>
    </Fragment>
  );
};
