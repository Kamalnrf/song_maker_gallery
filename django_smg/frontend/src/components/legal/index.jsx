/**
 * terms of service
 */

import React, { useState, Fragment } from "react";
import styled from "styled-components";

import { Portal } from "../common/portal";

import { TosText } from "./terms_of_service";
import { PrivacyText } from "./privacy_policy";
export { TosPage } from "./terms_of_service";
export { PrivacyPage } from "./privacy_policy";

/* Force <a> styling on anchor tags without href */
const A = styled.a`
  color: blue;
  font-weight: bold;
  text-decoration: underline;

  &:hover {
    cursor: pointer;
  }
`;

export const Tos = () => {
  const [showModal, setShowModal] = useState(false);
  const toggleModal = () => setShowModal(!showModal);
  return (
    <Fragment>
      {showModal ? (
        <Portal>
          <TosText onOk={toggleModal} />
        </Portal>
      ) : null}
      <A onClick={toggleModal}>Terms of Service</A>
    </Fragment>
  );
};

export const Privacy = () => {
  const [showModal, setShowModal] = useState(false);
  const toggleModal = () => setShowModal(!showModal);
  return (
    <Fragment>
      {showModal ? (
        <Portal>
          <PrivacyText onOk={toggleModal} />
        </Portal>
      ) : null}
      <A onClick={toggleModal}>Privacy Policy</A>
    </Fragment>
  );
};
