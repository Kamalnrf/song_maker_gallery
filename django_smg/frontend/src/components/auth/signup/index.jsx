import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";

import { register, clearError } from "../../../actions/auth.action";
import styled, {
  P,
  Button,
  Label,
  Checkbox,
  Description,
} from "../../common/styles";
import { ErrorArray } from "../../common/custom_error";
import { Tos, Privacy } from "../../legal";

import { ValidationMessages } from "./validation_messages";

export const PASSWD_MIN_LENGTH = 9;

// subtle adjust to align text with checkbox
const SpanCbAligned = styled.span`
  position: relative;
  bottom: 8px;
`;

const Header = styled.h1`
  margin: 0rem;
  font-size: 60px;
`;

const FlexForm = styled.form`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
  gap: 3rem;
`;

const SubmitButton = styled(Button)`
  font-size: 24px;
  height: auto;
  line-height: 40px;
  padding: 10px 5px 10px 5px;
  background-color: lightseagreen;
`;

const signup = (props) => {
  const [emailInput, updateEmail] = useState("");
  const [usernameInput, updateUsername] = useState("");
  const [passwordInput, updatePassword] = useState("");
  const [passwordConfirm, updateConfirm] = useState("");
  const [blanket, setBlanket] = useState(null);
  const clearBlanket = () => setBlanket(null);
  const [TOS, setTOS] = useState(false);

  useEffect(() => {
    document.title = props.title;
  }, []);

  /**
   * Validate form and inject a modal warning into "blanket" if it is not
   * valid. Checks for the following:
   *
   * - All fields are filled
   * - Password and confirm password are the same
   * - Terms of service box has been checked
   *
   * @returns {bool} whether form is valid
   */
  const validate = () => {
    if (
      emailInput === "" ||
      usernameInput === "" ||
      passwordInput === "" ||
      passwordConfirm === ""
    ) {
      setBlanket(
        <ErrorArray
          header="Blank Fields"
          message={["Required fields are blank", "All fields are required."]}
          onOk={clearBlanket}
        />
      );
      return false;
    }
    if (passwordInput !== passwordConfirm) {
      setBlanket(
        <ErrorArray
          header="Passwords do not match"
          message={[""]}
          onOk={clearBlanket}
        />
      );
      return false;
    }
    if (!TOS) {
      setBlanket(
        <ErrorArray
          header="Terms of Service"
          message={["You must accept the terms of service to make an account"]}
          onOk={clearBlanket}
        />
      );
      return false;
    }
    return true;
  };

  const submit = (e) => {
    e.preventDefault();
    if (validate()) {
      props.register({
        email: emailInput,
        username: usernameInput,
        password: passwordInput,
      });
    }
  };

  if (props.isAuthenticated) {
    return <Redirect to="/teacher" />;
  } else if (props.authError && !blanket) {
    const errors = Object.keys(props.authError).map(
      (k) => `${k}: ${props.authError[k]}`
    );
    setBlanket(
      <ErrorArray
        onOk={() => {
          props.clearError();
          clearBlanket();
        }}
        header="Validation Error"
        message={errors}
      />
    );
  }

  return (
    <div>
      {blanket}
      <Header>sign up!</Header>
      <Description>
        <FlexForm data-testid="signupForm" onSubmit={submit}>
          <div>
            <Label htmlFor="email">Email</Label>
            <input
              id="email"
              data-testid="emailInput"
              onChange={(event) => updateEmail(event.target.value)}
            />
            <Label htmlFor="username">Username</Label>
            <input
              id="username"
              data-testid="usernameInput"
              style={
                usernameInput.includes(" ") ? { borderColor: "red" } : null
              }
              value={usernameInput}
              onChange={(event) => updateUsername(event.target.value)}
            />
          </div>
          <div>
            <Label htmlFor="password">Password</Label>
            <input
              id="password"
              data-testid="passwordInput"
              type="password"
              value={passwordInput}
              onChange={(event) => updatePassword(event.target.value)}
              style={
                passwordInput &&
                passwordConfirm &&
                passwordInput !== passwordConfirm
                  ? { borderColor: "red" }
                  : null
              }
            />
            <Label htmlFor="confirm password">Confirm Password</Label>
            <input
              id="confirm password"
              data-testid="passwordConfirmInput"
              value={passwordConfirm}
              onChange={(e) => updateConfirm(e.target.value)}
              type="password"
              style={
                passwordInput &&
                passwordConfirm &&
                passwordInput !== passwordConfirm
                  ? { borderColor: "red" }
                  : null
              }
            />
          </div>
          <div>
            <ValidationMessages
              usernameInput={usernameInput}
              passwordInput={passwordInput}
              passwordConfirm={passwordConfirm}
            />
            <P>
              <SpanCbAligned>
                I agree to the <Tos /> and <Privacy />
                {"    "}
              </SpanCbAligned>
              <Checkbox
                id="tos"
                onClick={() => setTOS(!TOS)}
                data-testid="tosCheckbox"
              />
            </P>
            <div>
              <SubmitButton as="input" type="submit" value="Sign Up" />
            </div>
            <div>
              <Link to="/login">
                <Button>Already have an account? Login here!</Button>
              </Link>
            </div>
          </div>
        </FlexForm>
      </Description>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    authError: state.auth.authError,
    isAuthenticated: state.auth.isAuthenticated,
  };
};

export default connect(mapStateToProps, { register, clearError })(signup);
