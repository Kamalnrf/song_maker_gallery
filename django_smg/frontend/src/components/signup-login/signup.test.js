import React from "react";
import renderer from "react-test-renderer";
import {
  render,
  screen,
  fireEvent,
  act,
  waitFor,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import { Context } from "../../test/app_context";

import Signup from "./signup";
import { REGISTER } from "../../actions/types";
import { register } from "../../actions/auth.action";
import { Tos, Privacy } from "../legal";
import { ErrorArray } from "../generics/custom_error";

jest.mock("../generics/custom_error");
jest.mock("../legal");
jest.mock("../../actions/auth.action");

Tos.mockImplementation(() => <span>TOS MOCK</span>);
Privacy.mockImplementation(() => <span>PRIVACY POLICY MOCK</span>);
ErrorArray.mockImplementation(() => <span>MOCK ERRORARRAY</span>);

beforeEach(() => {
  render(
    <Context>
      <Signup />
    </Context>
  );
});

afterEach(() => {
  register.mockClear();
  ErrorArray.mockClear();
});

const submitForm = async (
  data = {
    email: "test@testuser.com",
    username: "testuser",
    password: "123456789",
  }
) => {
  await act(async () => {
    fireEvent.change(screen.getByTestId("emailInput"), {
      target: { value: data.email },
    });
    fireEvent.change(screen.getByTestId("usernameInput"), {
      target: { value: data.username },
    });
    fireEvent.change(screen.getByTestId("passwordInput"), {
      target: { value: data.password },
    });
    fireEvent.change(screen.getByTestId("passwordConfirmInput"), {
      target: { value: data.password },
    });
    fireEvent.click(screen.getByTestId("tosCheckbox"));
  });
  await act(async () => {
    fireEvent.click(screen.getByTestId("submit"));
  });
  return data;
};

describe("<Signup />", () => {
  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <Context>
          <Signup />
        </Context>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("tells user when passwd is too short", () => {
    expect(
      screen.getByText("Your password must be at least eight characters long")
    ).toBeVisible();
  });

  it("tells user when passswd is long enough with green text", async (done) => {
    // action: input password
    act(() => {
      fireEvent.change(screen.getByTestId("passwordInput"), {
        target: { value: "123456789" },
      });
    });

    // assertions
    await waitFor(() => {
      expect(
        screen.queryByText("Your password is at least eight characters long")
      ).toBeVisible();
      expect(
        screen.queryByText("Your password is at least eight characters long")
      ).toHaveStyle("color: green;");
    });
    done();
  });
  it("restores original passwd length message after deleting password", async (done) => {
    // actions
    // input password
    act(() => {
      fireEvent.change(screen.getByTestId("passwordInput"), {
        target: { value: "123456789" },
      });
    });
    // partially remove password
    await act(async () => {
      fireEvent.change(screen.getByTestId("passwordInput"), {
        target: { value: "123456" },
      });
    });

    // assertions
    await waitFor(() => {
      expect(
        screen.getByText("Your password must be at least eight characters long")
      ).toBeVisible();
    });
    done();
  });
  it("disallows empty fields", async (done) => {
    await act(async () => {
      fireEvent.click(screen.getByTestId("submit"));
    });
    await waitFor(() => {
      expect(ErrorArray).toHaveBeenCalledWith(
        {
          header: "Blank Fields",
          message: ["Required fields are blank", "All fields are required."],
          onOk: expect.anything(),
        },
        expect.anything() // other args from react framework, etc.
      );
    });
    done();
  });
  it("calls user register action on form submission", async (done) => {
    const data = await submitForm();
    await waitFor(() => {
      expect(register).toHaveBeenCalledTimes(1);
      expect(register).toHaveBeenCalledWith(data);
    });
    done();
  });
  it("renders 400 error messages from server", async (done) => {
    const fakeApiResponse = {
      username: ["A user with that username already exists"],
      email: ["Invalid email address"],
    };
    register.mockImplementation(() => (dispatch) => {
      dispatch({
        type: REGISTER,
        payload: {
          isAuthenticated: false,
          authError: fakeApiResponse,
          token: null,
          user: null,
        },
      });
    });
    await submitForm();
    expect(ErrorArray).toHaveBeenCalledWith(
      {
        header: "Validation Error",
        message: expect.arrayContaining([
          `username: ${fakeApiResponse["username"]}`,
          `email: ${fakeApiResponse["email"]}`,
        ]),
        onOk: expect.anything(),
      },
      expect.anything()
    );
    done();
  });
  it("normalizes authErrors that are strings before passing to <ErrorArray />", async (done) => {
    register.mockImplementation(() => (dispatch) => {
      dispatch({
        type: REGISTER,
        payload: {
          isAuthenticated: false,
          authError: "something bad happened",
          token: null,
          user: null,
        },
      });
    });
    await submitForm();
    expect(ErrorArray).toHaveBeenCalledWith(
      {
        message: expect.arrayContaining(["Error: something bad happened"]),
        header: "Validation Error",
        onOk: expect.anything(),
      },
      expect.anything()
    );
    done();
  });
  it("notifies user of 500 error response", () => {});
});
