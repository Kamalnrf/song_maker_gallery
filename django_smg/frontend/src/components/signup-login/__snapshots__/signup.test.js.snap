// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> matches snapshot 1`] = `
<div>
  <h1
    className="sc-bdfBwQ kmRABv"
  >
    sign up!
  </h1>
  <div
    className="styles__Description-emvDpe hIumWL"
  >
    <div
      className="sc-gsTCUz cCQKqw"
    >
      <h3>
        Email
      </h3>
      <input
        data-testid="emailInput"
        onChange={[Function]}
        onKeyDown={[Function]}
      />
      <h3>
        Username
      </h3>
      <input
        data-testid="usernameInput"
        onChange={[Function]}
        onKeyDown={[Function]}
        style={null}
        value=""
      />
    </div>
    <div
      className="sc-gsTCUz cCQKqw"
    >
      <h3>
        Password
      </h3>
      <input
        data-testid="passwordInput"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="password"
        value=""
      />
      <h3>
        Confirm Password
      </h3>
      <input
        data-testid="passwordConfirmInput"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="password"
        value=""
      />
    </div>
    <p
      className="styles__P-dIaYsU fltiwN"
    >
      Your password must be at least eight characters long
    </p>
    <span
      style={
        Object {
          "bottom": "8px",
          "position": "relative",
        }
      }
    >
      I agree to the 
      <span>
        TOS MOCK
      </span>
       and 
      <span>
        PRIVACY POLICY MOCK
      </span>
    </span>
    <input
      className="styles__Input-lbDPrm fzsTeW"
      data-testid="tosCheckbox"
      id="tos"
      onClick={[Function]}
      type="checkbox"
    />
    <button
      className="styles__Button-dyjLXS sc-hKgILt LXcou cgUWKs"
      data-testid="submit"
      onClick={[Function]}
    >
      Sign Up
    </button>
    <a
      href="/login"
      onClick={[Function]}
    >
      <button
        className="styles__Button-dyjLXS LXcou"
      >
        Already have an account? Login here!
      </button>
    </a>
  </div>
</div>
`;
